signingworker/worker.py

```diff
@@ -0,0 +1,187 @@
+import os
+import random
+import shutil
+import tempfile
+import urlparse
+import arrow
+from jsonschema import ValidationError
+from kombu import Exchange, Queue
+from kombu.mixins import ConsumerMixin
+import redo
+import logging
+import requests
+import sh
+import json
+import taskcluster
+
+from signingworker.task import validate_task
+from signingworker.exceptions import TaskVerificationError, \
+    ChecksumMismatchError, SigningServerError
+from signingworker.utils import my_ip, get_hash
+
+log = logging.getLogger(__name__)
+
+
+class SigningConsumer(ConsumerMixin):
+
+    def __init__(self, connection, exchange, queue_name, worker_type,
+                 taskcluster_config, signing_servers, signing_server_auth,
+                 supported_signing_scopes, tools_checkout):
+        self.connection = connection
+        self.exchange = Exchange(exchange, type='topic', passive=True)
+        self.queue_name = queue_name
+        self.worker_type = worker_type
+        self.routing_key = "*.*.*.*.*.*.{}.#".format(self.worker_type)
+        self.tc_queue = taskcluster.Queue(taskcluster_config)
+        self.signing_servers = signing_servers
+        # TODO: Explicitly set formats in manifest
+        self.signing_formats = ["mar", "gpg"]
+        self.signing_server_auth = signing_server_auth
+        self.supported_signing_scopes = supported_signing_scopes
+        self.tools_checkout = tools_checkout
+        self.cert = os.path.join(self.tools_checkout,
+                                 "release/signing/host.cert")
+
+    def get_consumers(self, consumer_cls, channel):
+        queue = Queue(name=self.queue_name, exchange=self.exchange,
+                      routing_key=self.routing_key, durable=True,
+                      exclusive=False, auto_delete=False)
+        return [consumer_cls(queues=[queue], callbacks=[self.process_message])]
+
+    def process_message(self, body, message):
+        task_id = None
+        run_id = None
+        try:
+            task_id = body["status"]["taskId"]
+            run_id = body["status"]["runs"][-1]["runId"]
+            log.debug("Claiming task %s, run %s", task_id, run_id)
+            self.tc_queue.claimTask(
+                task_id, run_id,
+                {"workerGroup": self.worker_type, "workerId": self.worker_type}
```

Comment:
Should workerId be the same for all instances of this, or unique per instance? I'm not really sure what it's used for in Taskcluster...